cpp/include/graph.hpp

```diff
@@ -240,10 +245,291 @@ class GraphCSC: public GraphCompressedSparseBase<VT,ET,WT> {
+    GraphCompressedSparseBaseView<VT,ET,WT>(offsets_, indices_, edge_data_, number_of_vertices_, number_of_edges_)
+  {}
+};
+
+
+/**
+ * @brief      TODO : Change this Take ownership of the provided graph arrays in COO format
+ *
+ * @param  source_indices        This array of size E (number of edges) contains the index of the source for each edge.
+ *                               Indices must be in the range [0, V-1].
+ * @param  destination_indices   This array of size E (number of edges) contains the index of the destination for each edge.
+ *                               Indices must be in the range [0, V-1].
+ * @param  edge_data             This array size E (number of edges) contains the weight for each edge.  This array can be null
+ *                               in which case the graph is considered unweighted.
+ * @param  number_of_vertices    The number of vertices in the graph
+ * @param  number_of_edges       The number of edges in the graph
+ */
+template <typename VT, typename ET, typename WT>
+struct GraphCOOContents {
+  VT number_of_vertices;
+  ET number_of_edges;
+  std::unique_ptr<rmm::device_buffer> src_indices;
+  std::unique_ptr<rmm::device_buffer> dst_indices;
+  std::unique_ptr<rmm::device_buffer> edge_data;
+};
+
+/**
+ * @brief       A constructed graph stored in COO (COOrdinate) format.
+ *
+ * This class will src_indices and dst_indicies (until moved)
+ *
+ * @tparam VT   Type of vertex id
+ * @tparam ET   Type of edge id
+ * @tparam WT   Type of weight
+ */
+template <typename VT, typename ET, typename WT>
+class GraphCOO {
+  VT number_of_vertices_;
+  ET number_of_edges_;
+  rmm::device_buffer src_indices_{};   ///< rowInd
+  rmm::device_buffer dst_indices_{};   ///< colInd
+  rmm::device_buffer edge_data_{}; ///< CSR data
+
+public:
+
+  /**
+   * @brief      Take ownership of the provided graph arrays in COO format
+   *
+   * @param  source_indices        This array of size E (number of edges) contains the index of the source for each edge.
+   *                               Indices must be in the range [0, V-1].
+   * @param  destination_indices   This array of size E (number of edges) contains the index of the destination for each edge.
+   *                               Indices must be in the range [0, V-1].
+   * @param  edge_data             This array size E (number of edges) contains the weight for each edge.  This array can be null
+   *                               in which case the graph is considered unweighted.
+   * @param  number_of_vertices    The number of vertices in the graph
+   * @param  number_of_edges       The number of edges in the graph
+   */
+  GraphCOO(VT number_of_vertices,
+           ET number_of_edges,
+           bool has_data = false):
+    number_of_vertices_(number_of_vertices),
+    number_of_edges_(number_of_edges),
+    src_indices_(sizeof(VT)*number_of_edges),
+    dst_indices_(sizeof(VT)*number_of_edges),
+    edge_data_(has_data? sizeof(WT)*number_of_edges : 0)
```

Comment:
Edge data will just be an empty array?

Reply:
Yes, if the graph does not have any weights associated with it, then the edge_data buffer is effectively not allocated.

Reply:
In https://github.com/rapidsai/cugraph/pull/799/files#diff-ba59ce02957abb53c35f1a1929fc0374R248,
trailing _ is used for constructor input parameters, and here, trailing _ is used for class member variables. This can be confusing and I think we'd better be consistent. And I think a more common approach across RAPIDS is to use preceding _ for class member variables (raft does this https://github.com/rapidsai/raft/pull/3/files#diff-245ee3425e4d2e84ecf153f459ec06efR154 and I think we'd better follow raft's convention).

Reply:
@seunghwak I am perfectly fine altering this. I was apprehensive keeping underscore because of this quote.

"Don't overlegislate naming, but do use a consistent naming convention: There are only two must-dos: a) never use "underhanded names," ones that begin with an underscore or that contain a double underscore;" (p2 , C++ Coding Standards, Herb Sutter and Andrei Alexandrescu)

Reply:
That is interesting, and did they just say not use use a name "begin" with an underscore? (but OK with a name end with an underscore, or disapprove both?)

Google C++ coding style guide asks to use a trailing underscore for class member variables (https://google.github.io/styleguide/cppguide.html#General_Naming_Rules), and I am curious whether a trailing underscore is OK but a preceding one is not, or Goolge and Herb & Andrei have different opinions.

Reply:
I found

https://www.drdobbs.com/c-coding-standards/184401863
"name private member variables likeThis_;"

and it seems like Herb & Andrei support using trailing _ but does not like preceding _. Do you know why? If there is a persuasive reason, we should better go for trailing _, but I can't think of any reason that trailing _ is better than preceding _.

Reply:
> I found
> 
> https://www.drdobbs.com/c-coding-standards/184401863
> "name private member variables likeThis_;"
> 
> and it seems like Herb & Andrei support using trailing _ but does not like preceding _. Do you know why? If there is a persuasive reason, we should better go for trailing _, but I can't think of any reason that trailing _ is better than preceding _.

[This](https://stackoverflow.com/a/20496955) could be a reason why.

Reply:
Thanks for the explanation!!! Then, we should consistently use trailing _, I will try to enforce the same for raft as well.

Reply:
I would have liked to make the same changes to the view classes but I thought it is out of scope for the current PR. Would like to hear your and @ChuckHastings 's thoughts.

Reply:
I think we need to settle on a few C++ coding standards (probably beyond just naming, perhaps we start with naming, white space vice tabs, indentation rules) and agree on them as a team.  Then I think we should come up with a strategy to update the code base to that agreed upon standard.  Perhaps something like:
* all new code will be developed to those standards
* code substantially changed in a PR will be adapted to these standards - but not code that is tangentially modified in a PR

Then we can get comfortable with our standards.  At some point, perhaps over a couple of releases, we could create issues to address updating older code to the new standards if it hasn't already been addressed in PRs based on the above bullets.

It might be best to coordinate this with the automatic code reformatting that has been proposed for a while but not yet implemented.

Toward this, I would consider the edits to the view classes tangential.  While I wouldn't oppose making the edits in this type of PR, I would not feel that they were necessary.

Reply:
> Toward this, I would consider the edits to the view classes tangential. While I wouldn't oppose making the edits in this type of PR, I would not feel that they were necessary.

Cleary they are not critical, but I just wanted to move in a direction to make our codebase more consistent than the opposite. If it takes too much work and we are under pressure, we can leave those as technical debt, but if we have room, I think it is a good idea to address what we can address now rather than pushing everything as future work (even if this can cause some additional work in the future).